APE_Black_Formatting.rst

```diff
@@ -0,0 +1,286 @@
+settings, that is, 88 characters per line and double quotes. Note that, by
+design, most of Black_ is inflexible, and there is not a way to pick-and-choose
+some elements and not others of Black's format.
+
+Implementing this change requires:
+
+1. Updating the `coding style`_:
+
+   * Adding documentation about Black, with a particular emphasis on the rationale expressed here and the exception policy described above.
+   * Adding documentation about using pre-commit, ala `pandas pre-commit`_.
+   * Document the Black exception guidelines (as described in this APE)
+   * Removing other references to code formatting in the documentation.
+
+2. Updating flake8 configuration to be compatible with Black.
+
+   This is straightforward and well documented. It may even be removed if it
+   does not provide any added checks that ``black –check`` does not.
+
+3. Checking Black in CI.
+
+   * Add a CI check that runs ``black --check`` on the whole code base, failing
+     if it meets a violation
+   * Allow for the Black CI bot to fix formatting errors via documented single
+     line comments. This is an opt-in step which could be used by maintainers to
+     fix errors prior to merging. Having such an option available could lower
+     the barrier for new contributors.
+
+4. Reformatting Astropy's code.  This will be done on a subpackage-by-subpackage
+   basis over a transition period. The process for each subpackage to be
+   transitioned is as follows:
+
+   * The sub package maintainer(s) create a branch (e.g.
+     ``blackify-coordinates``) to do the re-formatting.
+   * Edit ``pyproject.toml`` to remove the black exclusion for the sub-package.
+   * Run ``black --skip-string-normalization astropy/<sub-package>`` on the
+     sub-package directory.  Commit the changes black makes and make a draft
+     pull request.
+   * The maintainer(s) should review all changes and highlight for discussion
+     any  changes that are judged to be valid exceptions (see above for what
+     "valid exception" means). These can be reverted by copying-and-pasting
+     from the "changes" interfaces to the local copy and applying the
+     appropriate Black directives to prevent auto-formatting. Commit and push
+     these changes for review.
+   * Once the PR submitter and reviewer are both satisfied with the changes
+     then squash the commit(s) down to a single commit and force push.
+   * Run ``black astropy/<sub-package>`` to apply just the string normalization
+     stage and commit and push.
+   * Edit .git-blame-ignore-revs to include the commits comprising PR to this
+     point.
+   * Convert the draft PR to a normal PR (ready for review). Once it passes CI
+     then it can be merged without further review.
+
+A sub-package will not have Black checked in CI until it has completed these
+steps.
+
+Backward compatibility
+----------------------
+
+This APE renders all previous code formatting styles no longer compatible. In
```

Comment:
Not clear what 'compatible' means - maybe better said as 'This APE would superseded any previous code formatting styles recommended for the core package'

Reply:
the enforcing of pre-commit bot, without black could be compatible, and therefore I feel it's a mistake to merge these two proposals together and sell the advantages of one as the advantage of the other and a reasoning to live with the disadvantages. (e.g. pre-commit without black but enforcing linting has the huge potential to be the middle ground)

Reply:
As far as pre-commit and the pre-commit bot are concerned, the infrastructure team has been moving forward with using them in astropy independent of this APE. The introduction of pre-commit has already occurred, while this APE was being drafted it was realized that pre-commit was not documented for astropy developers. The use of the bot is a natural extension of using pre-commit which should help ease the burden of any current (or future) linting or style checks on contributions.